rust/datafusion/src/execution/aggregate.rs

```diff
@@ -347,6 +346,73 @@ impl AggregateFunction for SumFunction {
     }
 }
 
+/// Implementation of AVG aggregate function
+#[derive(Debug)]
+struct AvgFunction {
+    data_type: DataType,
+    sum_value: SumFunction,
+    count_value: CountFunction,
+}
+
+impl AvgFunction {
+    fn new(data_type: &DataType) -> Self {
+        Self {
+            data_type: DataType::Float64,
+            sum_value: SumFunction::new(data_type),
+            count_value: CountFunction::new(),
+        }
+    }
+}
+
+impl AggregateFunction for AvgFunction {
+    fn name(&self) -> &str {
+        "avg"
+    }
+
+    fn accumulate_scalar(&mut self, value: &Option<ScalarValue>) -> Result<()> {
+        self.sum_value.accumulate_scalar(value)?;
+        self.count_value.accumulate_scalar(value)?;
+
+        Ok(())
+    }
+
+    fn accumulate_batch(&mut self, array: ArrayRef) -> Result<()> {
+        self.sum_value.accumulate_batch(array.clone())?;
+        self.count_value.accumulate_batch(array)?;
+
+        Ok(())
+    }
+
+    fn result(&self) -> Option<ScalarValue> {
+        let sum = match self.sum_value.result() {
+            Some(ScalarValue::UInt8(a)) => a as f64,
+            Some(ScalarValue::UInt16(a)) => a as f64,
+            Some(ScalarValue::UInt32(a)) => a as f64,
+            Some(ScalarValue::UInt64(a)) => a as f64,
+            Some(ScalarValue::Int8(a)) => a as f64,
+            Some(ScalarValue::Int16(a)) => a as f64,
+            Some(ScalarValue::Int32(a)) => a as f64,
+            Some(ScalarValue::Int64(a)) => a as f64,
+            Some(ScalarValue::Float32(a)) => a as f64,
+            Some(ScalarValue::Float64(a)) => a as f64,
+            _ => {
+                return None;
+            }
+        };
+        let count = match self.count_value.result() {
+            Some(ScalarValue::UInt64(a)) => a as f64,
+            _ => {
```

Comment:
Nop, all `ScalerValue` except `UInt64` won't be accepted. The `_` is mandatory here.

Reply:
Right but I mean it's better to be explicit about what is acceptable.  `None` is a valid variant but any other variant except `ScalarVaue::UInt64` is an error.

Reply:
You are right, but function `result` returns an `Option`. Do we need to change its signature to return a `Result` in order to represent a error case?

Reply:
Yea, I was going to say just panic, but I'm not sure that's the best either....  Let's see what Andy thinks.  In any case, this can be addressed (if at all) in a follow up PR.